source journal's homepage_url from OpenAlex

smierz opened the pull request:
This PR sources homepage_urls for journals from OpenAlex:
* querying OpenAlex stays close to the existing code in `crossref` with light changes
  *  removed the email parameter from the requesting method as it is never used
  * always taking the value from config + loading the config only once at startup
  * because config is located in config file, I need pytest-mock to set value in tests
* simple auto-generated db migration, adding column "`homepage_url`" to the `JournalBase` object
* when querying OpenAlex, I'm only putting one of the ISSNs into the query parameters to avoid repetition
* copy-pasted the code to update a journal's homepage_url from  `crossref.store_papers` because I'm not familiar with SQLAlchemy

sneakers-the-rat commented on src/paper_feeds/services/journal_service.py:
```diff
@@ -0,0 +1,45 @@
+            if journal:
+                journal.homepage_url = homepage_url
+                session.add(journal)
+                session.commit()
```
I think since we're not returning anything here we can just commit at the end and forego refreshing?

smierz replied:
you're right, that was also my first version but something didn't work but then later it turned out it was not the code here but the alembic migration and I forgot to change this code back :/

sneakers-the-rat commented on src/paper_feeds/app.py:
```diff
@@ -41,13 +41,18 @@ async def index(request: Request):
 
 @app.post('/search')
-async def search(request: Request, search: Annotated[str, Form()]):
+async def search(request: Request,
```
we will probably want to separate this out from crossref service as we add more data providers, but that's not for this PR :)

sneakers-the-rat commented on src/paper_feeds/services/journal_service.py:
```diff
@@ -0,0 +1,45 @@
+    issns_to_query = []
+    for journal in journals:
+        if journal.issn:
+            issns_to_query += [journal.issn[0].value]
```
note to self, we should add a way of doing 'preferred issn' and getting the web ISSN if available, i have a hunch that the assumption that metadata will be equal across ISSNs (so we can treat them as arbitrary) will prove incorrect, but for now this is fine!

sneakers-the-rat commented on src/paper_feeds/services/journal_service.py:
```diff
@@ -0,0 +1,45 @@
+
+
+def get_journals_homepage(journals: list[Journal]):
+    # extract one issn of each journal newly added to DB
```
would be nice to move this into a docstring, eventually we will have docs!

smierz replied:
added docstrings to all methods in `openalex`, `journal_service` and `journal_repository`

sneakers-the-rat commented on src/templates/pages/journal.html:
```diff
@@ -11,7 +11,7 @@ <h2 class="journal-title">{{ journal.title }}</h2>
         <div class="metadata-item">
             <span class="metadata-key">Homepage:</span>
-            <span class="metadata-value">TODO add homepage to journal model</span>
+            <span class="metadata-value">{{ journal.homepage_url }}</span>
```
nice thank you :)

sneakers-the-rat commented on src/templates/pages/journal.html:
```diff
@@ -11,7 +11,7 @@ <h2 class="journal-title">{{ journal.title }}</h2>
         </div>
         <div class="metadata-item">
             <span class="metadata-key">Homepage:</span>
-            <span class="metadata-value">TODO add homepage to journal model</span>
+            <span class="metadata-value">{{ journal.homepage_url }}</span>
         </div>
```
(this should probably be an `<a>` tag and wrapped in `if` if we don't have it sometimes. i can never remember the jinja strip whitespace params so something like this, but not important to get it exactly right this PR since we will have to do style changes later anyway)

```suggestion
        {% if journal.homepage_url %}
        <div class="metadata-item">
            <span class="metadata-key">Homepage:</span>
            <a href="{{ journal.homepage_url }}">
                <span class="metadata-value">{{ journal.homepage_url }}</span>
            </a>
        </div>
        {% endif %}
```

smierz replied:
you're right, this way it would be cleaner

smierz replied:
fixed it

sneakers-the-rat commented on src/paper_feeds/app.py:
```diff
@@ -58,6 +63,8 @@ async def search(request: Request, search: Annotated[str, Form()]):
 async def journal_page(request: Request, issn:str):
     journal = crossref.load_journal(issn)
+    if journal.homepage_url is None:
+        journal.homepage_url = '-'
```
re: my comment below making the field optional in presentation, no strong preference between displaying null value like this vs. just not displaying at all - we should make these endpoints return JSON when expected so this isn't a problem in case someone wants computer-readable access to this API at some point.

sneakers-the-rat commented on tests/test_openalex.py:
```diff
@@ -0,0 +1,50 @@
+)
+def test_get_journal_homepages(issn, homepage):
+    issn_homepage_map = get_journal_homepages([issn])
+    assert issn_homepage_map[issn] == homepage
```
one additional test would be great here just to ensure that the `update_journals_homepage_urls` function works and correctly writes to the db :). I think that would just be something like

```python
@pytest.mark.parametrize(
    'issn,homepage',
    ...
)
def test_update_journal_homepages(issn, homepage):
    update_journals_homepage_urls([issn])
    journal = load_journal(issn)
    assert journal.homepage_url == homepage

```

smierz replied:
separated business logic from db operations (`journal_service` vs `journal_repository`) and added tests for both methods in `journal_repository`

sneakers-the-rat commented on src/paper_feeds/services/openalex.py:
```diff
@@ -0,0 +1,51 @@
+
+    # minimize API calls using approach outlined in OpenAlex blog
+    # https://blog.ourresearch.org/fetch-multiple-dois-in-one-openalex-api-request/
+    slice_size = 50
```
should this be hardcoded like this or any reason to make this a parameter, eg. if we want to chain this to some other updating call and control paging size all together?

smierz replied:
50 is the upper limit for chaining values via "OR" operator ([link](https://docs.openalex.org/how-to-use-the-api/get-lists-of-entities/filter-entity-lists#addition-or)).  
--> using slices of 50 entities reduces the amount of requests to the API
--> page size needs to be set accordingly, otherwise defaults to 25 and have to additionally paginate the resulting two pages

sneakers-the-rat reviewed (commented):
Hell ya, this mostly looks good to me, thank you!!!! some questions/comments in there before merging, but all minor ones. 

re: config - the config uses [`pydantic-settings`](https://docs.pydantic.dev/latest/concepts/pydantic_settings/) and is configured to read from environment variables with the prefix `JOURNALFEEDS_` or from a `.env` file (see `.env.example`), see here for an example on how to use that in tests: https://github.com/sneakers-the-rat/paper-feeds/blob/e30221267dc6ff5a4c8d397dd8b34b1e23267ea1/tests/test_crossref.py#L42

No preference on whether the `get_openalex` function is private (and thus wouldn't receive config as an optional argument) or not, but i do have a mild preference for treating the `Config` object as a singleton that gets instantiated when used or passed explicitly, rather than created as a module-level object - the goal is to have config be single-source and so we want to strongly discourage storing and modifying it. is there reason to switch that pattern aside from needing to be able to mock it?

If you're not familiar with SQLAlchemy I can write the test to test the DB update, and we can also punt testing whether we need `commit` in the loop or not until we do actual profiling/perf updates :)

smierz commented:
thanks for the quick review, I will look into it asap, but probably due to work not before the weekend :/

smierz commented:
update:
* separated all db operations from `journal_service` into separate file `journal_repository`
* added tests for methods in `journal_repository` (test coverage up to 79%)
* added docstrings to all methods in `openalex`, `journal_service` and `journal_repository`

smierz commented on src/paper_feeds/config.py:
```diff
@@ -43,3 +43,6 @@ def reload(self) -> bool:
         self.log_dir.mkdir(exist_ok=True, parents=True)
+
+
+config = Config()
```
kicked it out in favor of dependency injection

smierz commented:
re:config

> i do have a mild preference for treating the Config object as a singleton that gets instantiated when used or passed explicitly, rather than created as a module-level object - the goal is to have config be single-source and so we want to strongly discourage storing and modifying it.

hmm I think we were both going for a singleton approach, but the difference is the passing of parameters:
I was going for imports (=import singleton config from config.py) while you would prefer dependency injection (pass parameters from config singleton in app.py)

Both are valid approaches for handling config propagation in applications and I'm ok with either.
The only concern I have is with creating new Config objects around the application (like `crossref.crossref_get` does) because I do  not want the config to be mutable from outside the application (= environment variables) during runtime (good luck finding this "bug" if user accidentally changes them).

Anyway, I kicked out the config variable from config.py and switched to dependency injection passing the email parameter through to `openalex._openalex_get`

sneakers-the-rat commented:
Sry for the delay, will merge when I get to keyboard. I think ill deploy this today too

sneakers-the-rat commented:
thanks for this! looks great :)

sneakers-the-rat commented:
Released in v0.0.5 :) 
https://github.com/sneakers-the-rat/paper-feeds/releases/tag/v0.0.5
https://github.com/sneakers-the-rat/paper-feeds/tree/v0.0.5
https://pypi.org/project/paper-feeds/0.0.5/
